docs/separate-benchmark-module.md

```diff
@@ -0,0 +1,102 @@
+# Creating Separate Module for Benchmarks
+
+This guide will walk you through the process of creating a separate module for benchmarks in your Kotlin project. 
+Creating a separate module for benchmarks is especially beneficial when you are integrating benchmarks into an existing project. 
+Here are a few advantages of doing so:
+
+1. **Organization**: It helps maintain a clean and organized project structure. Segregating benchmarks from the main code makes it easier to navigate and locate specific code segments.
+
+2. **Isolation**: Separating benchmarks ensures that the benchmarking code does not interfere with your main code or test code. This isolation guarantees accurate measurements without unintentional side effects.
+
+3. **Flexibility**: Creating a separate module allows you to manage your benchmarking code independently. You can compile, test, and run benchmarks without impacting your main source code.
+
+## Step-by-step Setup Guide
+
+Below are the step-by-step instructions to set up a separate module for benchmarks in both Kotlin Multiplatform and Kotlin/JVM projects:
+
+### Kotlin Multiplatform Project
+
+Follow these steps to set up a separate compilation for benchmarks:
+
+1. **Define New Compilation**:
+
+    Start by creating a new compilation in your target of choice (e.g. jvm, js, native, wasm etc.). 
+    In this example, we're associating the new compilation `benchmark` with the `main` compilation of the `jvm` target. 
+    This association allows the benchmark compilation to access the internal API of the main compilation, 
+    which is particularly useful when benchmarks need to measure the performance of specific components 
+    or functionalities within the main codebase.
+
+    ```kotlin
+    // build.gradle.kts
+    kotlin { 
+        jvm { 
+            compilations.create('benchmark') {
+                associateWith(compilations.main)
+            }
+        }
+    }
+    ```
+
+2. **Register Benchmark Compilation**:
+
+    Conclude by registering your new benchmark compilation using its default source set name. 
+    In this instance, `jvmBenchmark` is the name of the default source set of the `benchmark` compilation.
+
+    ```kotlin
+    // build.gradle.kts
+    benchmark {
+        targets {
+            register("jvmBenchmark")
+        }
+    }
+    ```
+
+For additional information, refer to the [Kotlin documentation on creating a custom compilation](https://kotlinlang.org/docs/multiplatform-configure-compilations.html#create-a-custom-compilation).
+and the [documentation on associating compiler tasks](https://kotlinlang.org/docs/gradle-configure-project.html#associate-compiler-tasks).
+[Here is a sample Kotlin Multiplatform project](/examples/kotlin-multiplatform) with a separate compilation for benchmarks.
+
+### Kotlin/JVM Project
+
+Set up a separate benchmark source set by following these simple steps:
```

Comment:
According the title, the document is about creating a separate module for benchmarks, but this sections describes how to add a benchmarks source set instead.

For me as a reader it raises a few questions:
1) Can I create a separate module w/ benchmarks for K/JVM project? Or it's only for KMP and in JVM-only project I can only add benchmarks into a separate source set?
2) If I can add a separate module into K/JVM project too, do I always need a main source set and separate benchmarks source set?

Reply:
`kotlinx-benchmark` supports both Kotlin Multiplatform and Kotlin/JVM projects.

In Kotlin Multiplatform (K/M) project:
* If the benchmark code is integrated into the main compilation of Kotlin targets (for instance, when the project is specifically dedicated to benchmarks), Kotlin targets should be registered as benchmark targets. This is documented in the updated README.
* If one prefers to separate benchmark code from the main compilation, separate compilations should be created and registered as benchmark targets, using the names of their main source set. The steps for this are outlined in the guide.

In Kotlin/JVM project:
* If the benchmark code is part of the main source set (for example, in a project dedicated to benchmarks), the main source set should be registered as the benchmark target. A separate guide will be provided for this scenario.
* If one wishes to segregate benchmark code from the main source set, a separate source set should be created and registered as the benchmark target, as described in this guide.

It's important to note that in K/M projects, compilations are used, while in K/JVM projects, source sets are employed, as there is no separate Kotlin compilation in K/JVM projects. The term "module" is used in this guide to encompass both "compilation" and "source set" contexts. I would appreciate any suggestions to improve the wording.

Reply:
The described setup might not be the most logical, but that is how the library works. See [BenchmarksExtension.kt](https://github.com/Kotlin/kotlinx-benchmark/blob/3f32f2d9d9732a2a01d231134c52e6554428b519/plugin/main/src/kotlinx/benchmark/gradle/BenchmarksExtension.kt#L42) for more details on how `kotlinx-benchmark` handles the registration of benchmark targets.

Reply:
I guess the main source of the confusion is

> The term "module" is used in this guide to encompass both "compilation" and "source set" contexts. I would appreciate any suggestions to improve the wording.

Idea uses the term "module" as a synonym for Gradle sub-project, so when I hear "module" the only thing I'm thinking of is sub-projects. 

The term "module" has multiple meaning and the document does not specify that here the word "module" means a compilation or a source set, depending on project type. 
As the word "module" is used only a couple of times and referred in the main README, perhaps it's worth to abstain from using it at all and instead write something like "separating benchmark sources from the main project sources"?

Then, the preamble could be rewritten to something like:
```
This guide will walk you through the process of separating benchmarks from the main sources in your Kotlin project. 
That is especially beneficial when you are integrating benchmarks into an existing project.
 ...
 <the list of benefits>
 ...
Depending on a project type (Multiplatform or JVM), different approaches should be employed to keep benchmark sources aside of the main project sources. The guide below will walk you through required steps for each of the supported project types.
```

I'm not going to claim that it's the best wording, but it seems to be less ambiguous both in terms of the intention and the approach required to achieve it.

Reply:
I've changed the guide to use "source set" throughout. The reason is eventually the benchmark code in placed into source sets in both Kotlin Multiplatform and Kotlin/JVM projects. Please take a look if the new wording is any good?

Reply:
Thanks for addressing the issues. Everything looks good to me!